timeutil/schedule.go

```diff
@@ -0,0 +1,234 @@
+// (different day) is the *not* same interval
+//
+func (sched *Schedule) SameInterval(t1, t2 time.Time) bool {
+	if !sched.Matches(t1) || !sched.Matches(t2) {
+		// FIXME: or return error here?
+		return false
+	}
+
+	if sched.Weekday != "" {
+		t1Year, t1Week := t1.ISOWeek()
+		t2Year, t2Week := t2.ISOWeek()
+		return t1Year == t2Year && t1Week == t2Week
+	}
+
+	return t1.Year() == t2.Year() && t1.Month() == t2.Month() && t1.Day() == t2.Day()
+}
+
+var weekdayMap = map[string]int{
+	"sun": 0, "sunday": 0,
+	"mon": 1, "monday": 1,
+	"tue": 2, "tuesday": 2,
+	"wed": 3, "wednesday": 3,
+	"thu": 4, "thursday": 4,
+	"fri": 5, "friday": 5,
+	"sat": 6, "saturday": 6,
+}
+
+// parseWeekday gets an input like "mon@9:00-11:00" or "9:00-11:00"
+// and extracts the weekday of that schedule string (which can be
+// empty). It returns the remainder of the string, the weekday
+// and an error.
+func parseWeekday(s string) (rest, weekday string, err error) {
+	if !strings.Contains(s, "@") {
+		return s, "", nil
+	}
+
+	l := strings.SplitN(s, "@", 2)
+	weekday = strings.ToLower(l[0])
+	_, ok := weekdayMap[weekday]
+	if !ok {
+		return "", "", fmt.Errorf("cannot parse %q: not a valid day", l[0])
+	}
+	rest = l[1]
+
+	return rest, weekday, nil
+}
+
+// parseTimeInterval gets an input like "9:00-11:00"
+// and extracts the start and end of that schedule string and
+// returns them and any errors.
+func parseTimeInterval(s string) (start, end TimeOfDay, err error) {
+	if strings.Contains(s, "@") {
+		return start, end, fmt.Errorf("cannot parse %q: contains invalid @", s)
+	}
+	l := strings.SplitN(s, "-", 2)
+	if len(l) != 2 {
+		return start, end, fmt.Errorf("cannot parse %q: not a valid interval", s)
+	}
+
+	start, err = ParseTime(l[0])
+	if err != nil {
+		return start, end, fmt.Errorf("cannot parse %q: not a valid time", l[0])
+	}
+	end, err = ParseTime(l[1])
+	if err != nil {
+		return start, end, fmt.Errorf("cannot parse %q: not a valid time", l[1])
+	}
+	if end.Less(start) {
+		return start, end, fmt.Errorf("cannot parse %q: not a valid interval", s)
+	}
+
+	return start, end, nil
+}
+
+// parseSingleSchedule parses a schedule string like "mon@9:00-11:00" or
+// "9:00-11:00" and returns a Schedule struct and an error.
+func parseSingleSchedule(s string) (*Schedule, error) {
+	rest, weekday, err := parseWeekday(s)
+	if err != nil {
+		return nil, err
+	}
+	start, end, err := parseTimeInterval(rest)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Schedule{
+		Weekday: weekday,
+		Start:   start,
+		End:     end,
+	}, nil
+}
+
+// ParseSchedule takes a schedule string in the form of:
+//
+// 9:00-15:00 (every day between 9am and 3pm)
+// 9:00-15:00,21:00-22:00 (every day between 9am,5pm and 9pm,10pm)
+// thu@9:00-15:00 (only Thursday between 9am and 3pm)
+// fri@9:00-11:00,mon@13:00-15:00 (only Friday between 9am and 3pm and Monday between 1pm and 3pm)
+// fri@9:00-11:00,13:00-15:00  (only Friday between 9am and 3pm and every day between 1pm and 3pm)
```

Comment:
There are some missing pieces which I suppose we're leaving behind in the name of doing it sooner, but there are also some incompatibilities in terms of what I thought we had agreed to.

For completeness, this is the full example:
```
tue
tue,thu
tue-thu
9:00
9:00,15:00
9:00-15:00
tue,thu@9:00-15:00
tue@9:00;thu@15:00
mon,wed-fri@9:00-11:00,13:00-15:00
```
So, in comparison we:

1. lack of individual times implying a +1h window
2. lack of weekday ranges
3. lack of multiple weekdays associating to particular times and mulitple times associated to particular weekday (ranges or not)
4. are interpreting `;` and `,` differently (oops!)

Points 3 and 4 are related, and are an incompatibility rather than just something missing. Under what I thought we were implementing, that would mean "Friday between (9am and 11am, 1pm and 3pm)" instead of "(Friday between 9am and 11am), (1pm and 3pm)".

In particular, the last line in the original example meant: "(Monday and Wednesday and Thursday and Friday) between (9am and 11am, 1pm and 3pm)".

Taking this chance, I also suggest replacing ";" in the original syntax by "/", so that we don't need to quote in the shell.